// mo-ccs/index.js
var fs = require("fs");
var path = require("path");
var moUtils = require("mo-utils");
var moImg = require("mo-img");
var async = require('async');

var templateDir = path.join(__dirname, "template");

/**
 * 新建cca工程。
 * @param dir
 * @param ccaName
 */
exports.newCca = function(dir, ccaName){
    var projDir = path.join(dir, ccaName);
    if(fs.existsSync(projDir)) {
        console.warn("项目【%s】已经存在，不能重复创建！", projDir);
        return false;
    }
    moUtils.mkdirSync(projDir);
    moUtils.mkdirSync(path.join(projDir, "Json"));
    moUtils.mkdirSync(path.join(projDir, "Resources"));

    var ccaJsonStr = fs.readFileSync(path.join(templateDir, "cca.json")).toString();
    fs.writeFileSync(path.join(projDir, "Json", "cca.json"), ccaJsonStr);

    var ccaXmlAniStr = fs.readFileSync(path.join(templateDir, "cca.xml.animation")).toString();
    fs.writeFileSync(path.join(projDir, "cca.xml.animation"), ccaXmlAniStr);
    return true;
};
/**
 * 将cca重置成项目需要用的解构。
 * @param srcCcaPath
 * @param dstCcaPath
 */
exports.resetCca = function(srcCcaPath, dstCcaPath){
    moUtils.mkdirSync(dstCcaPath);
    moUtils.mkdirSync(path.join(dstCcaPath, "Json"));
    moUtils.mkdirSync(path.join(dstCcaPath, "Resources"));

    var jsonFiles = fs.readdirSync(path.join(srcCcaPath, "Json"));
    var jsonFile;
    for(var i = 0, li = jsonFiles.length; i < li; i++){
        var itemi = jsonFiles[i];
        var extname = path.extname(itemi);
        if(extname && extname.toLowerCase() == ".json") {
            jsonFile = itemi;
            break;
        }
    }
    var resPath = path.join(srcCcaPath, "Resources");
    var results = {};

    moUtils.walkDirSync(resPath, [".jpg", ".png"], function(filePath){
        var basename = path.basename(filePath).toLowerCase();
        if(results[basename]) console.warn("资源文件发生重名：【%s】", filePath);
        results[basename] = true;
        moUtils.copyFileSync(filePath, path.join(dstCcaPath, "Resources", basename));
    });

    var jsonPath = path.join(srcCcaPath, "Json", jsonFile);
    var jsonData = require(jsonPath);
    fs.writeFileSync(path.join(dstCcaPath, "Json/cca.json"), JSON.stringify(resetJsonData(jsonData)));

    var ccaXmlAniStr = fs.readFileSync(path.join(templateDir, "cca.xml.animation")).toString();
    fs.writeFileSync(path.join(dstCcaPath, "cca.xml.animation"), ccaXmlAniStr);
};

function resetJsonData(jsonData, ccaName){
    resetArmatureData(jsonData.armature_data[0], ccaName);
    resetAnimationData(jsonData.animation_data[0], ccaName);
    resetTextureData(jsonData.texture_data, ccaName);
    return jsonData;
}
function resetArmatureData(armatureData, ccaName){
    armatureData.name = ccaName || "cca";
    var boneData = armatureData.bone_data;
    if(!boneData) return;
    for(var i = 0, li = boneData.length; i < li; i++){
        var bd = boneData[i];
        if(!bd) continue;
        var displayData = bd.display_data;
        if(!displayData) continue;
        for(var j = 0, lj = displayData.length; j < lj; ++j){
            var dd = displayData[j];
            if(ccaName) dd.name = ccaName + "/" + dd.name;
            else dd.name = path.basename(dd.name);
        }
    }
}
function resetAnimationData(animationData, ccaName){
    animationData.name = ccaName || "cca";
}
function resetTextureData(textureData, ccaName){
    if(!textureData) return;
    for(var i = 0, li = textureData.length; i < li; i++){
        var itemi = textureData[i];
        if(ccaName) itemi.name = ccaName + "/" + itemi.name;
        else itemi.name = path.basename(itemi.name);
    }
}


/**
 * 发布单独的cca文件
 * @param ccaPath
 * @param scale
 * @param {{outDir:String, outName:String, uglifyJsonOpt:{JSON_KEY:Object, JSON_DEFAULT_VALUE:Object}}} opt
 * @param cb
 * @returns {*}
 */
exports.publishCca = function(ccaPath, scale, opt, cb){
    var ccaName = path.basename(ccaPath);//获取cca工程名称
    scale = scale || 1;//缩放
    opt = opt || {};//选项
    var outDir = opt.outDir || path.join(ccaPath, "output");
    var outName = opt.outName || ccaName;//导出的名字
    var __tempDir = path.join(ccaPath, "__temp");
    moImg.pack(path.join(ccaPath, "Resources"), scale, outDir, outName, __tempDir, function(err){
        if(err) return cb(err);
        //导出json内容
        var jsonPath;
        moUtils.walkDirSync(path.join(ccaPath, "Json"), ".json", function(filePath){
            jsonPath = filePath;
        });
        var jsonData = resetJsonData(require(jsonPath), outName);
        jsonData.config_file_path = [outName + ".plist"];
        if(opt.uglifyJsonOpt) jsonData = moUtils.uglifyJson(jsonData, opt.uglifyJsonOpt);
        fs.writeFile(path.join(outDir, outName + ".exportjson"), JSON.stringify(jsonData, null, 0), cb);
    });
}

//exports.publishCca(path.join(__dirname, "../../__test/monster_dashuyao"), 0.5, {outName : "ttt"});

function isPreviewNode(jsonData){
    if(jsonData.classname == "ImageView") {
        if(jsonData.options.fileNameData.path.match(/效果图/g)) return true;
    }
    return false;
}

function delPreviewNode(parentNode){
    var children = parentNode.children || [];
    for(var i = 0; i < children.length;){
        var child = children[i];
        if(isPreviewNode(child)){
            children.splice(i, 1);
        }else{
            delPreviewNode(child);
            i++;
        }
    }
}

/**
 * 发布成单独的ccui的ExportJson文件
 * @param ccuiJsonPath
 * @param {{outDir:String, outName:String, ignoredTextures:Array, uglifyJsonOpt:{JSON_KEY:Object, JSON_DEFAULT_VALUE:Object}}} opt
 * @param cb
 * @returns {*}
 */
exports.publishCcuiJson = function(ccuiJsonPath, opt, cb){
    var ccuiName = path.basename(ccuiJsonPath);
    opt = opt || {};//选项
    opt.ignoredTextures = opt.ignoredTextures || [];
    var outDir = opt.outDir || path.join(path.join(ccuiJsonPath, "../../output"));
    var oName = opt.outName || ccuiName;//导出的名字

    moUtils.mkdirSync(outDir);

    var jsonData = require(ccuiJsonPath);
    delPreviewNode(jsonData.widgetTree);
    var jsonStr = JSON.stringify(jsonData, null, 0);
    var result = jsonStr.match(/"path"[\s]*:[\s]*"[^\r\n"]+"/g);
    var texArr = [];
    if(result){
        for (var i = 0, li = result.length; i < li; i++) {
            var rStr = result[i];
            var str = rStr.match(/[^\r\n"]+\//);
            if(str){
                str = str[0].substring(0, str[0].length-1);
                if(opt.ignoredTextures.indexOf(str) < 0 && texArr.indexOf(str + ".plist") < 0){
                    texArr.push(str + ".plist")
                }
            }
        }
    }
    jsonData.textures = texArr;
    jsonStr = JSON.stringify(jsonData, null, 2);
    jsonStr = jsonStr.replace(/"useMergedTexture"[\s]*:[\s]*false/g, '"useMergedTexture": true');
    jsonStr = jsonStr.replace(/"resourceType"[\s]*:[\s]*0/g, '"resourceType": 1');

    //重新生成json对象
    jsonData = eval('(' + jsonStr + ')');
    if(opt.uglifyJsonOpt) jsonData = moUtils.uglifyJson(jsonData, opt.uglifyJsonOpt);
    fs.writeFile(path.join(outDir, moUtils.changeExtname(oName, ".exportjson")), JSON.stringify(jsonData, null, 0), cb);
}
//exports.publishCcuiJson(path.join(__dirname, "../../__test/ccui/Json/uiHeroCard2.json"));

/**
 * 打包ccui中的模块图片资源城.plist和.png
 * @param ccuiImgModuleName
 * @param cfg
 * @param cb
 * @returns {*}
 */
exports.publishCcuiImg = function(ccuiDir, imgName, scale, opt, cb){
    scale = scale || 1;//缩放
    opt = opt || {};//选项
    var outDir = opt.outDir || path.join(ccuiDir, "output");
    var __tempDir = path.join(ccuiDir, "__temp");
    moImg.pack(path.join(ccuiDir, "Resources", imgName), scale, outDir, imgName, __tempDir, cb);
}

//exports.publishCcuiImg(path.join(__dirname, "../../__test/ccui/"), "ui_common", 0.5, null);

/**
 * 发布ccui资源。
 * @param ccuiDir
 * @param scale
 * @param opt
 * @param cb
 */
exports.publishCcui = function(ccuiDir, scale, opt, cb){
    scale = scale || 1;//缩放
    opt = opt || {};//选项
    var outDir = opt.outDir || path.join(ccuiDir, "output");
    moUtils.walkDirOneByOne(path.join(ccuiDir, "Json"), ".json", function(filePath, cb1){
        var ccuiJsonOpt = {
            outDir : outDir,
            ignoredTextures : opt.ignoredTextures,
            uglifyJsonOpt : opt.uglifyJsonOpt
        };
        exports.publishCcuiJson(filePath, ccuiJsonOpt, cb1);
    }, function(err){
        if(err) return cb(err);
        var files = fs.readdirSync(path.join(ccuiDir, "Resources"));
        async.mapLimit(files, 1, function(imgName, cb1){
            var ccuiImgOpt = {
                outDir : outDir
            }
            exports.publishCcuiImg(ccuiDir, imgName, scale, ccuiImgOpt, cb1);
        }, cb);
    })
};
//exports.publishCcui(path.join(__dirname, "../../__test/ccui/"), 0.5);
